src/cmd/tools/dtest/docker/harness/a_common.go

```diff
@@ -0,0 +1,231 @@
+	if o.overrideDefaults {
+		return o
+	}
+
+	if len(o.source) == 0 {
+		o.source = defaultOpts.source
+	}
+
+	if len(o.containerName) == 0 {
+		o.containerName = defaultOpts.containerName
+	}
+
+	if len(o.dockerFile) == 0 {
+		o.dockerFile = defaultOpts.dockerFile
+	}
+
+	if len(o.portList) == 0 {
+		o.portList = defaultOpts.portList
+	}
+
+	if len(o.mounts) == 0 {
+		o.mounts = defaultOpts.mounts
+	}
+
+	if o.iOpts == nil {
+		o.iOpts = defaultOpts.iOpts
+	}
+
+	return o
+}
+
+func newOptions(name string) *dockertest.RunOptions {
+	return &dockertest.RunOptions{
+		Name:      name,
+		NetworkID: networkName,
+	}
+}
+
+func setupNetwork(pool *dockertest.Pool) error {
+	networks, err := pool.Client.ListNetworks()
+	if err != nil {
+		return err
+	}
+
+	for _, n := range networks {
+		if n.Name == networkName {
+			if err := pool.Client.RemoveNetwork(networkName); err != nil {
+				return err
+			}
+
+			break
+		}
+	}
+
+	_, err = pool.Client.CreateNetwork(dc.CreateNetworkOptions{Name: networkName})
+	return err
+}
+
+func setupVolume(pool *dockertest.Pool) error {
+	volumes, err := pool.Client.ListVolumes(dc.ListVolumesOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, v := range volumes {
+		if volumeName == v.Name {
+			if err := pool.Client.RemoveVolume(volumeName); err != nil {
+				return err
+			}
+
+			break
+		}
+	}
+
+	_, err = pool.Client.CreateVolume(dc.CreateVolumeOptions{
+		Name: volumeName,
+	})
+
+	return err
+}
+
+func exposePorts(
+	opts *dockertest.RunOptions,
+	portList []int,
+) *dockertest.RunOptions {
+	ports := make(map[dc.Port][]dc.PortBinding, len(portList))
+	for _, p := range portList {
+		port := fmt.Sprintf("%d", p)
+
+		portRepresentation := dc.Port(fmt.Sprintf("%s/tcp", port))
+		binding := dc.PortBinding{HostIP: "0.0.0.0", HostPort: port}
+		entry, found := ports[portRepresentation]
+		if !found {
+			entry = []dc.PortBinding{binding}
+		} else {
+			entry = append(entry, binding)
+		}
+
+		ports[portRepresentation] = entry
+	}
+
+	opts.PortBindings = ports
+	return opts
+}
+
+func setupMount(dest string) string {
+	src := os.TempDir()
+	src = "/Users/arnikola/go/src/github.com/m3db/m3/scripts/" +
+		"docker-integration-tests/cold_writes_simple"
+	return fmt.Sprintf("%s:%s", src, dest)
+}
+
+func toResponse(
+	resp *http.Response,
+	response proto.Message,
+	logger *zap.Logger,
+) error {
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		logger.Error("could not read body", zap.Error(err))
+		return err
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode/100 != 2 {
+		logger.Error("status code not 2xx",
+			zap.Int("status code", resp.StatusCode),
+			zap.String("status", resp.Status))
+		return fmt.Errorf("status code %d", resp.StatusCode)
+	}
+
+	err = jsonpb.Unmarshal(bytes.NewReader(b), response)
+	if err != nil {
+		logger.Error("unable to unmarshal response",
+			zap.Error(err),
+			zap.Any("response", response))
+		return err
+	}
+
+	return nil
+}
+
+func toResponseThrift(
```

Comment:
Should we perhaps be using native TChannel+Thrift client instead of HTTP+JSON with the thrift endpoints?

Following methods let us use the native TChannel+Thrift client in integration tests:
https://github.com/m3db/m3/blob/master/src/dbnode/integration/client.go#L44-L105